velox/common/caching/CachedFactory.h

```diff
@@ -208,135 +121,34 @@ class CachedFactory {
-    std::unique_ptr<Key> key,
-    std::mutex* mu)
-    : wasCached_(wasCached),
-      value_(value),
-      cache_(cache),
-      key_(std::move(key)),
-      mu_(mu) {}
-
-template <typename Key, typename Value, typename Comparator, typename Hash>
-CachedPtr<Key, Value, Comparator, Hash>::~CachedPtr() {
-  clear();
-}
-
-template <typename Key, typename Value, typename Comparator, typename Hash>
-CachedPtr<Key, Value, Comparator, Hash>::CachedPtr(CachedPtr&& other) {
-  wasCached_ = other.wasCached_;
-  value_ = other.value_;
-  key_ = std::move(other.key_);
-  cache_ = other.cache_;
-  mu_ = other.mu_;
-  other.value_ = nullptr;
-}
-
-template <typename Key, typename Value, typename Comparator, typename Hash>
-CachedPtr<Key, Value, Comparator, Hash>&
-CachedPtr<Key, Value, Comparator, Hash>::operator=(CachedPtr&& other) {
-  clear();
-  wasCached_ = other.wasCached_;
-  value_ = other.value_;
-  key_ = std::move(other.key_);
-  cache_ = other.cache_;
-  mu_ = other.mu_;
-  other.value_ = nullptr;
-  return *this;
-}
-
-template <typename Key, typename Value, typename Comparator, typename Hash>
-void CachedPtr<Key, Value, Comparator, Hash>::clear() {
-  if (!value_)
-    return;
-  if (cache_) {
-    if (mu_) {
-      std::lock_guard<std::mutex> l(*mu_);
-      cache_->release(*key_);
-    } else {
-      cache_->release(*key_);
-    }
-  } else {
-    delete value_;
-  }
-}
-
-template <
-    typename Key,
-    typename Value,
-    typename Generator,
-    typename Sizer,
-    typename Comparator,
-    typename Hash>
-CachedPtr<Key, Value, Comparator, Hash>
-CachedFactory<Key, Value, Generator, Sizer, Comparator, Hash>::generate(
+template <typename Key, typename Value, typename Generator>
+std::pair<bool, Value> CachedFactory<Key, Value, Generator>::generate(
     const Key& key) {
   std::unique_lock<std::mutex> pending_lock(pendingMu_);
   {
     std::lock_guard<std::mutex> cache_lock(cacheMu_);
-    Value* value = cache_->get(key);
+    auto value = cache_->get(key);
     if (value) {
-      return CachedPtr<Key, Value, Comparator, Hash>(
-          /*wasCached=*/true,
-          value,
-          cache_.get(),
-          std::make_unique<Key>(key),
-          &cacheMu_);
+      return std::make_pair(true, value.value());
     }
   }
+
   if (pending_.contains(key)) {
     pendingCv_.wait(pending_lock, [&]() { return !pending_.contains(key); });
     // Will normally hit the cache now.
     {
       std::lock_guard<std::mutex> cache_lock(cacheMu_);
-      Value* value = cache_->get(key);
+      auto value = cache_->get(key);
       if (value) {
-        return CachedPtr<Key, Value, Comparator, Hash>(
-            /*wasCached=*/false,
-            value,
-            cache_.get(),
-            std::make_unique<Key>(key),
-            &cacheMu_);
+        return std::make_pair(true, value.value());
       }
     }
     pending_lock.unlock();
     return generate(key); // Regenerate in the edge case.
   } else {
```

Comment:
nit: can we simplify the control flow here:
```
if (condition A) {
  return xxx;
}

condition B

instead of 

if (condition A) {
  ...
  return;
} else {
  condition B.
  ...
}
```